src/js/app/file.js

```diff
@@ -24,22 +24,75 @@ export default class File extends Base {
 
   /**
    * Content of the file after being decoded
-   * @returns {string}
+   * @returns {Promise<string>}
    */
-  content(encoding = 'utf8') {
-    const decoder = new TextDecoder(encoding);
-    return decoder.decode(new DataView(this._rawContent));
-  }
+  content(...args) { return Spaces.getFileContentProvider().content(this, ...args); }
 
   /**
    * Set the contents of this file
+   * @param {Promise<ArrayBuffer>} content - Promise which returns the file content
    * @param {cont} - the content of the file as bytes
    */
-  setRawContent(cont) { this._rawContent = cont; }
+  setRawContent(...args) { return Spaces.getFileContentProvider().setRawContent(this, ...args); }
 
-  rawContent() { return this._rawContent; }
+  /**
+   * Gets the raw content of a file
+   * @returns {Promise<ArrayBuffer>}
+   */
+  rawContent(...args) { return Spaces.getFileContentProvider().rawContent(this, ...args); }
+
+  /**
+   * Gets the size of the raw content
+   * @returns {number}
+   */
+  rawContentSize(...args) { return Spaces.getFileContentProvider().rawContentSize(this, ...args); }
 
-  rawContentSize() { return this._rawContent ? this._rawContent.byteLength : 0; }
+  /**
+   * Returns whether this file has raw content
+   * @returns {boolean}
+   */
+  hasRawContent(...args) { return Spaces.getFileContentProvider().hasRawContent(this, ...args); }
+
+  /**
+   * Clears raw content from the cache. Not guaranteed to clear content, cache may ignore.
+   * Should be used to ensure we don't sent content in memory to the main thread.
+   */
+  clearRawContent(...args) { return Spaces.getFileContentProvider().clearRawContent(this, ...args); }
+
+
+  // Returned from Spaces.getContentProvider as get
+  static __content(file, encoding = 'utf8') {
+    if (file.hasRawContent()) {
+      const decoder = new TextDecoder(encoding);
+      return file.rawContent()
+        .then((content) => {
+          if (!content) {
+            return Promise.reject(new Error('content undefined, rawContent must return an ArrayBuffer'));
+          }
+          return decoder.decode(new DataView(content));
+        });
+    }
+    return Promise.reject(new Error('raw content does not exist'));
+  }
+
+  // Returned from Spaces.getContentProvider as set
+  static __setRawContent(file, cont) {
+    if (typeof cont.then !== 'function') {
+      return Promise.reject(new Error('content must be a promise'));
+    }
+    return cont.then((content) => {
+      file._rawContent = content;
+      return Promise.resolve(file);
```

Comment:
fyi, no need to return resolved promises, `return file;` would do the same thing